File: traffic-gen/scenarios/test/workflow-http-dns.lua
-- Equivalent to the Python/JSON example; no Lua executes in the packet path.
local tg=require('snowtg')
local peer=os.getenv('SNOWTG_PEER') or '192.168.10.234'
return tg.scenario('business-http-dns', {duration=5,cps=20,concurrency=16,seed=17,
    classes={tg.transaction('business',{vars={peer=peer},dataset='workflow-users.csv',steps={
        -- Copy response scalars into ctx; Header index=1 selects the second occurrence.
        tg.dns_step('resolve','${vars.peer}','alias.snowtg.test',15353,{extract={address=tg.extract('address')}}),
        tg.http_step('fetch','${ctx.address}',18080,{path='/chunked',keepalive=true,
            extract={token=tg.extract('json','/token'),['repeat']=tg.extract('header','x-repeat',{index=1})},
            checks={tg.check(tg.extract('status'),'==',200),tg.check(tg.extract('json','/items/0'),'==',7)}}),
        -- Only describe control flow here; native execution chooses the dataset row and wait.
        tg.think('pause',nil,{minimum=3,maximum=8}),
        tg.branch('route',tg.check(tg.ref('data.route'),'==','A'),{['then']='route_a',otherwise='route_b'}),
        -- The json filter quotes its value; next skips the other path and joins at finish.
        tg.http_step('route_a','${ctx.address}',18080,{path='/use',method='POST',keepalive=true,
            headers={['X-Route']='A'},body='{"token":${ctx.token|json},"user":${data.user|json}}',next='finish'}),
        tg.http_step('route_b','${ctx.address}',18080,{path='/use',method='POST',keepalive=true,
            headers={['X-Route']='B'},body='{"token":${ctx.token|json},"user":${data.user|json}}'}),
        tg.http_step('finish','${ctx.address}',18080,{path='/close'})}})},
    -- Separate whole-business acceptance from fetch-step latency and final socket drain.
    assertions={tg.assertion('success_rate','==',1),tg.assertion('success_rate','==',1,{step='fetch'}),
        tg.assertion('latency_ms','<',100,{quantile=.99,step='fetch'}),tg.assertion('drained_live_sockets','==',0)}})

File: traffic-gen/scenarios/test/workflow-http-dns.py
"""Business TPS: DNS -> JSON/Header capture -> think/branch -> POST -> close.
Run with snowtg.py run; SNOWTG_PEER points at test/workflow_peer.py's host.
Dataset rows are immutable and cycle by planned transaction ordinal.
"""
import os
from snowtg import (scenario, transaction, dns_step, http_step, think, branch,
                    extract, ref, check, assertion)
peer = os.environ.get('SNOWTG_PEER', '192.168.10.234')
plan = scenario('business-http-dns', duration=5, cps=20, concurrency=16, seed=17,
    classes=[transaction('business', vars={'peer': peer}, dataset='workflow-users.csv', steps=[
        # Save the resolved address in private ctx for later dynamic HTTP targets.
        dns_step('resolve', '${vars.peer}', 'alias.snowtg.test', 15353,
                 extract={'address': extract('address')}),
        http_step('fetch', '${ctx.address}', 18080, path='/chunked', keepalive=True,
                  extract={'token': extract('json', '/token'), 'repeat': extract('header', 'x-repeat', index=1)},
                  checks=[check(extract('status'), '==', 200), check(extract('json', '/items/0'), '==', 7)]),
        # Native seed/ordinal/step sampling controls this wait, not Python random or sleep.
        think('pause', minimum=3, maximum=8),
        branch('route', check(ref('data.route'), '==', 'A'), then='route_a', otherwise='route_b'),
        # The json filter includes quotes. Jump to finish to avoid executing route_b too.
        http_step('route_a', '${ctx.address}', 18080, path='/use', method='POST', keepalive=True,
                  headers={'X-Route': 'A'}, body='{"token":${ctx.token|json},"user":${data.user|json}}', next='finish'),
        http_step('route_b', '${ctx.address}', 18080, path='/use', method='POST', keepalive=True,
                  headers={'X-Route': 'B'}, body='{"token":${ctx.token|json},"user":${data.user|json}}'),
        http_step('finish', '${ctx.address}', 18080, path='/close')])],
    # Unscoped SLOs cover the business; fetch latency measures that network step only.
    assertions=[assertion('success_rate', '==', 1),
                assertion('success_rate', '==', 1, step='fetch'),
                assertion('latency_ms', '<', 100, quantile=.99, step='fetch'),
                assertion('drained_live_sockets', '==', 0)])
